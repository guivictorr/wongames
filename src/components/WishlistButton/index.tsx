import Button, { ButtonProps } from 'components/Button'
import Spinner from 'components/Spinner'
import { useWishlist } from 'hooks/useWishlist'
import { useSession } from 'next-auth/client'
import { useState } from 'react'
import { Favorite, FavoriteBorder } from 'styled-icons/material-outlined'

type WishlistButton = {
  id: string
  hasText?: boolean
} & Pick<ButtonProps, 'size'>

const WishlistButton = ({ id, hasText, size = 'small' }: WishlistButton) => {
  const [loading, setLoading] = useState(false)
  const { addToWishlist, isInWishlist, removeFromWishlist } = useWishlist()
  const [session] = useSession()
  const buttonLabel = isInWishlist(id)
    ? 'Remove from wishlist'
    : 'Add to wishlist'

  const toggleWishlist = async () => {
    setLoading(true)
    isInWishlist(id) ? await removeFromWishlist(id) : await addToWishlist(id)
    setLoading(false)
  }

  if (!session) return null

  return (
    <Button
      size={size}
      aria-label={buttonLabel}
      role="button"
      onClick={toggleWishlist}
      minimal
      icon={
        loading ? (
          <Spinner />
        ) : isInWishlist(id) ? (
          <Favorite />
        ) : (
          <FavoriteBorder />
        )
      }
    >
      {hasText && buttonLabel}
    </Button>
  )
}

export default WishlistButton
