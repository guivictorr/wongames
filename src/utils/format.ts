export const formatNumber = (
  value: number,
  options: Intl.NumberFormatOptions
) => {
  return new Intl.NumberFormat('en-US', options).format(value)
}
