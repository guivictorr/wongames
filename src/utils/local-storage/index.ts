const APP_KEY = 'WONGAMES'

export const getStorageItem = (key: string) => {
  if (typeof window === 'undefined') return

  const item = window.localStorage.getItem(`${APP_KEY}_${key}`)

  return JSON.parse(item!)
}

export const setStorageItem = (key: string, value: string | string[]) => {
  if (typeof window === 'undefined') return

  return window.localStorage.setItem(`${APP_KEY}_${key}`, JSON.stringify(value))
}
