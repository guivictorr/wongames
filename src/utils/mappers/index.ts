import { HighlightFragment } from 'graphql/generated/HighlightFragment'
import { QueryGames_games } from 'graphql/generated/QueryGames'
import { QueryHome_banners } from 'graphql/generated/QueryHome'

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map((banner) => ({
    title: banner.title,
    subtitle: banner.subtitle,
    img: `http://localhost:1337${banner.image?.url}`,
    ...(banner.button && {
      buttonLabel: banner.button.label,
      buttonLink: banner.button.link
    }),
    ...(banner.ribbon && {
      ribbon: banner.ribbon.title,
      ribbonColor: banner.ribbon.color,
      ribbonSize: banner.ribbon.size
    })
  }))
}

export const gameSliderMapper = (
  games: QueryGames_games[] | undefined | null
) => {
  return games
    ? games.map((game) => ({
        id: game.id,
        title: game.name,
        slug: game.slug,
        developer: game.developers[0].name,
        img: `http://localhost:1337${game.cover?.url}`,
        price: game.price
      }))
    : []
}

export const highlightMapper = (
  highlight: HighlightFragment | null | undefined
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background?.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage?.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment
      }
    : {}
}
