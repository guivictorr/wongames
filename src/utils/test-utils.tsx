import { ReactElement } from 'react'
import { ThemeProvider } from 'styled-components'
import { render, RenderOptions } from '@testing-library/react'
import {
  CartContextDefaultValues,
  CartContext,
  CartContextData
} from 'hooks/useCart'
import theme from 'styles/theme'

type CustomRenderProps = {
  cartProviderProps?: CartContextData
} & Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {}
) => {
  return render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        {ui}
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions
  )
}

export * from '@testing-library/react'
export { customRender as render }
