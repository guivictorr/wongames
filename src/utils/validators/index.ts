import { UsersPermissionsRegisterInput } from 'graphql/generated/globalTypes'
import Joi from 'joi'

const fieldValidators = {
  username: Joi.string().min(5).required(),
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required(),
  password: Joi.string().required(),
  confirm_password: Joi.string()
    .valid(Joi.ref('password'))
    .required()
    .messages({
      'any.only': 'confirm password does not match with password'
    })
}

export function signUpValidate(values: UsersPermissionsRegisterInput) {
  const schema = Joi.object(fieldValidators)
  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

export type FieldErrors = {
  [key: string]: string
}

function getFieldErrors(errorObject: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (errorObject.error) {
    errorObject.error.details.forEach((detail) => {
      errors[detail.path.join('.')] = detail.message
    })
  }

  return errors
}

type SignInValues = Omit<UsersPermissionsRegisterInput, 'username'>
export function signInValidate(values: SignInValues) {
  const { email, password } = fieldValidators

  const schema = Joi.object<SignInValues>({ email, password })
  const validation = schema.validate(values, { abortEarly: false })
  return getFieldErrors(validation)
}

type ForgotValidateParams = Pick<UsersPermissionsRegisterInput, 'email'>
export function forgotValidate(values: ForgotValidateParams) {
  const { email } = fieldValidators

  const schema = Joi.object({ email })
  const validation = schema.validate(values, { abortEarly: false })
  return getFieldErrors(validation)
}

type ResetValidateParams = {
  password: string
  confirm_password: string
}
export function resetValidate(values: ResetValidateParams) {
  const { password, confirm_password } = fieldValidators

  const schema = Joi.object({ password, confirm_password })
  const validation = schema.validate(values, { abortEarly: false })
  return getFieldErrors(validation)
}
