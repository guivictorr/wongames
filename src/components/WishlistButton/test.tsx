import userEvent from '@testing-library/user-event'
import {
  WishlistContextData,
  WishlistContextDefaultValues
} from 'hooks/useWishlist'
import { render, screen } from 'utils/test-utils'

import WishlistButton from '.'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useSession = jest.spyOn(require('next-auth/client'), 'useSession')

useSession.mockImplementation(() => [
  {
    jwt: '123',
    user: {
      email: 'email@gmail.com'
    }
  }
])

describe('<WishlistButton />', () => {
  it('should render a button to add to wishlist', () => {
    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => false
    }

    const { container } = render(<WishlistButton id="1" />, {
      wishlistProviderProps
    })

    expect(
      screen.getByRole('button', { name: 'Add to wishlist' })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a button to remove from wishlist', () => {
    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => true
    }

    render(<WishlistButton id="1" />, {
      wishlistProviderProps
    })

    expect(
      screen.getByRole('button', { name: 'Remove from wishlist' })
    ).toBeInTheDocument()
  })

  it('should render a button with add to wishlist text', () => {
    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => false
    }

    render(<WishlistButton id="1" hasText />, {
      wishlistProviderProps
    })

    expect(screen.getByText('Add to wishlist')).toBeInTheDocument()
  })

  it('should render a button with remove from wishlist text', () => {
    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => true
    }

    render(<WishlistButton id="1" hasText />, {
      wishlistProviderProps
    })

    expect(screen.getByText('Remove from wishlist')).toBeInTheDocument()
  })

  it('should not render if not logged', () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const useSession = jest.spyOn(require('next-auth/client'), 'useSession')

    useSession.mockImplementationOnce(() => [null])

    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => true
    }

    render(<WishlistButton id="1" hasText />, {
      wishlistProviderProps
    })

    expect(screen.queryByRole('button')).not.toBeInTheDocument()
  })

  it('should add item to the wishlist', () => {
    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => false,
      addToWishlist: jest.fn()
    }

    render(<WishlistButton id="1" hasText />, {
      wishlistProviderProps
    })

    userEvent.click(screen.getByRole('button', { name: 'Add to wishlist' }))

    expect(wishlistProviderProps.addToWishlist).toHaveBeenCalledWith('1')
  })

  it('should add item to the wishlist', () => {
    const wishlistProviderProps: WishlistContextData = {
      ...WishlistContextDefaultValues,
      isInWishlist: () => true,
      removeFromWishlist: jest.fn()
    }

    render(<WishlistButton id="1" hasText />, {
      wishlistProviderProps
    })

    userEvent.click(
      screen.getByRole('button', { name: 'Remove from wishlist' })
    )

    expect(wishlistProviderProps.removeFromWishlist).toHaveBeenCalledWith('1')
  })
})
