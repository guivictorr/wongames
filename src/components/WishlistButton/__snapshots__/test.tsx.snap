// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WishlistButton /> should render a button to add to wishlist 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: linear-gradient( 178.59deg,#ff5f5f -14.51%,#f062c0 102.86%,#f23131 102.86% );
  cursor: pointer;
  border-radius: 0.4rem;
  border: 0;
  padding: 0.8rem;
  height: 3rem;
  font-size: 1.2rem;
  background: none;
  color: #F231A5;
}

.c0:hover {
  background: none;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

.c0:hover {
  color: #e20e8d;
}

.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

<button
  aria-label="Add to wishlist"
  class="c0"
  role="button"
>
  <svg
    aria-hidden="true"
    class="c1"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
    />
  </svg>
</button>
`
